### ISSUE TITLE
Data Loss Mitigation Plan for Tempo 

### ISSUE THREAD

**Issue by kadhamecha-conga** [grafana/tempo#2609]
hi team,

can you please guide us on how can we reduce data loss from Tempo perspective?

there are two cases which we want to perform:

1.  Move from multi AZ to single AZ 
2.  Migrate from one storage class to another storage class 

can you please guide us on how we can minimize the data loss from tempo perspective?
what config changes we need to do in tempo?

please let us know , if you need more information.

thanks.

**Comment by joe-elliott** [comment]
Tempo is not aware of the AZ's or storage class of the objects it reads. Making these changes to your objects should have no impact on Tempo.

I would refer to your cloud provider's docs on how to make these updates.

**Comment by kadhamecha-conga** [comment]
hi @joe-elliott,

can you guide us on what config change requires, where ingester aggressively writing data to remote storage (s3 ) ?
to minimize data loss where we perform change to k8s objects  (like migrating pvc ) ?

thanks.

**Comment by joe-elliott** [comment]
if you're looking to make a change to the ingesters and would like to guarantee that all data is flushed before shutting it down you can either:

- [call the `/shutdown` endpoint](https://grafana.com/docs/tempo/latest/api_docs/#shutdown)
  this will remove the ingester from the ring so it no longer receives traffic and then flush all its data to object storage
- [set `flush_all_on_shutdown` to true](https://github.com/grafana/tempo/blob/main/docs/sources/tempo/configuration/_index.md#ingester)
  this will force the ingester to always completely flush it's disk on shutdown. note that k8s may force kill the ingester unless you configure the shutdown period.

is this what you're looking for?

**Comment by kadhamecha-conga** [comment]
hi @joe-elliott ,

we can consider this option to flush all data on pvc to backend storage.

i tried with suggested changes in ingester, but is not supporting in tempo, we are using tempo 2.0.0

```
    ingester:
      flush_all_on_shutdown: true
      lifecycler:
        ring:
          kvstore:
            store: memberlist
          replication_factor: 3
        tokens_file_path: /var/tempo/tokens.json
```
Error logs : 
```
failed parsing config: failed to parse configFile /conf/tempo.yaml: yaml: unmarshal errors:
          line 32: field flush_all_on_shutdown not found in type ingester.Config
```

is killing pod ( consider hpa scaling case ) is also flushing data on shutdown ?

thanks.

**Comment by joe-elliott** [comment]
> is killing pod ( consider hpa scaling case ) is also flushing data on shutdown ?

no. not unless `flush_all_on_shutdown` is set to true. the default case leaves data on disk that is not in object store.

> i tried with suggested changes in ingester, but is not supporting in tempo, we are using tempo 2.0.0

My bad. This feature is in main and will be included in Tempo 2.2. We are looking to cut this in the next month or so.

**Comment by kadhamecha-conga** [comment]
hi @joe-elliott  , 

what's behavior of flush_all_on_shutdown  ?

does it remove data from pvc and upload it to s3 ? and if data move from pvc to s3, how can we verify that it is successfully move to backend ?

thanks.

**Comment by joe-elliott** [comment]
> does it remove data from pvc and upload it to s3
yup

> and if data move from pvc to s3, how can we verify that it is successfully move to backend ?
you could mount the pvc and check that all completed blocks were in the backend and the wal was empty.

this can change based on config but the way we have it:
/var/tempo/wal <- wal block folders are in here
/var/tempo/wal/blocks <- complete blocks are in tenant folders here

**Comment by zhongchen** [comment]
```
 note that k8s may force kill the ingester unless you configure the shutdown period.
```
It seems to me that there is a bug in the code where the `flushOp` queued by enabling`flush_all_on_shutdown` gets abandoned immediately due to the `shutdown` signal terminating the `flush queue`.


![Screenshot 2023-08-08 at 2 39 41 PM](https://github.com/grafana/tempo/assets/3200761/51ebae1b-f1b8-49d5-84c9-56afdc1b3f8a)

In the screenshot, I can see the flushOp gets abandoned immediately.

**Comment by joe-elliott** [comment]
On shutdown the ingester should remove itself from the ring so it no longer receives traffic. It then waits for everything to flush and exits.

https://github.com/grafana/tempo/blob/0afe9a8db48dd5e4085747bb697cef9cc566e541/modules/ingester/ingester.go#L165

We only log the "Abandoning" line if the flush queues are stopped:

https://github.com/grafana/tempo/blob/0afe9a8db48dd5e4085747bb697cef9cc566e541/modules/ingester/flush.go#L248

It's possible for something like the following to happen:
- Receive SIGTERM
- Exit the ring 
- Flush everything correctly
- Stop the flush queue
- The ingester cuts the head block due to the timeout and the pictured message is logged. However the block should be empty b/c the ingester is not receiving data due to step #2.

But that should only happen if there's a considerable amount of time after stopping the flush queue. Can you review the disks and see if those head blocks have any traces in them?

**Comment by zhongchen** [comment]
@joe-elliott 
Looking at the code, I can see the flushQueue get sopped right after the shutdown flush is called.
https://github.com/grafana/tempo/blob/1968370615474a2d9bce3b336045d65dd63fa1c4/modules/ingester/ingester.go#L164-L172

When stopping a flushQueue, I don't see it waiting for any signal before stopping it. And in the screenshot I had in the previous comment, I noticed the timestamps of `queuing` and `abandoning` were the same, which made me wonder whether we wait for flush to complete before we stop the flushQueue.
https://github.com/grafana/tempo/blob/161ab9d4bb727ad0684aa3a600f024806a9e06c9/pkg/flushqueues/exclusivequeues.go#L72-L78

**Comment by zhongchen** [comment]
> Can you review the disks and see if those head blocks have any traces in them?
I am deploying Tempo in AWS EKS using the distributed-helm chart. I haven't looked into how to mount a persistent volume yet.

**Comment by joe-elliott** [comment]
> Looking at the code, I can see the flushQueue get sopped right after the shutdown flush is called.

Yes, but that `i.flushRemaining()` _should_ block until everything has flushed. discussion here:

https://github.com/grafana/tempo/blob/bc3f543bfcbac85e90d38faa4a34c0d495afd40f/modules/ingester/ingester.go#L150

it is a bit complex and certainly possible there's a hole in the logic.

> I noticed the timestamps of queuing and abandoning were the same, which made me wonder whether we wait for flush to complete before we stop the flushQueue.

It's a good question and certainly worth researching. Maybe we can signal the queues to stop taking new items, but continue to serve old items while the flush is occuring? it might shore up any issues with this logic.

**Comment by zhongchen** [comment]
Just needed the `PersistentVolume` for ingesters and verified that `flush_all_on_shutdown` was working as expected.

**Comment by joe-elliott** [comment]
To be clear, you're saying there is no bug and everything looks good from your side? if so i'd like to close this issue.

**Comment by zhongchen** [comment]
Yes.

**Comment by joe-elliott** [comment]
Awesome. Thanks for the debugging @zhongchen
